src/main/java/de/bonndan/nivio/input/ItemRelationProcessor.java

```diff
@@ -18,69 +18,129 @@ protected ItemRelationProcessor(ProcessLog processLog) {
     }
 
     @Override
-    public void process(LandscapeDescription input, Landscape landscape) {
-        input.getItemDescriptions().all().forEach(serviceDescription -> {
-            Item origin = landscape.getItems().pick(serviceDescription);
-            if (!input.isPartial()) {
-                processLog.debug(String.format("Clearing relations of %s", origin));
-                origin.getRelations().clear(); //delete all relations on full update
-            }
-        });
+    public ProcessingChangelog process(LandscapeDescription input, Landscape landscape) {
+
+        ProcessingChangelog changelog = new ProcessingChangelog();
+        List<Relation> processed = new ArrayList<>();
 
         input.getItemDescriptions().all().forEach(itemDescription -> {
             Item origin = landscape.getItems().pick(itemDescription);
+            List<Relation> affected = new ArrayList<>();
 
             for (RelationDescription relationDescription : itemDescription.getRelations()) {
-                Optional<Relation> update = update(relationDescription, landscape, origin);
-                update.ifPresent(relation -> {
-                    origin.getRelations().remove(relation);
-                    origin.getRelations().add(relation);
-                    if (relation.getSource() == origin) {
-                        relation.getTarget().getRelations().remove(relation);
-                        relation.getTarget().getRelations().add(relation);
-                    } else {
-                        relation.getSource().getRelations().remove(relation);
-                        relation.getSource().getRelations().add(relation);
+
+                if (!isValid(relationDescription, landscape)) {
+                    continue;
+                }
+
+                Optional<Relation> existing = getExisting(relationDescription, landscape, origin);
+                if (existing.isPresent()) {
+                    Relation update = update(relationDescription, existing.get());
+                    affected.add(update);
+                    processed.add(update);
+                    processLog.info(String.format("Updating relation between %s and %s", update.getSource(), update.getTarget()));
+                    List<String> changes = existing.get().getChanges(update);
+                    if (!changes.isEmpty()) {
+                        changelog.addEntry(update, ProcessingChangelog.ChangeType.UPDATED, String.join(";", changes));
                     }
-                });
+                    continue;
+                }
+
+                Relation created = create(relationDescription, landscape);
+                affected.add(created);
+                processed.add(created);
+                processLog.info(String.format(origin + ": Adding relation between %s and %s", created.getSource(), created.getTarget()));
+                changelog.addEntry(created, ProcessingChangelog.ChangeType.CREATED, null);
             }
+
```

Comment:
Also, what you want do do here is use the Optionals a little better, e.g. refactor to:
````
                Optional<Relation> optionalRelation = getRelation(relationDescription, landscape, origin);
                optionalRelation.ifPresentOrElse((relation) -> {
                        Relation update = update(relationDescription, relation);
                        affected.add(update);
                        processed.add(update);
                        processLog.info(String.format("Updating relation between %s and %s", update.getSource(), update.getTarget()));
                        List<String> changes = relation.getChanges(update);
                        if (!changes.isEmpty()) {
                            changelog.addEntry(update, ProcessingChangelog.ChangeType.UPDATED, String.join(";", changes));
                        }
                    }, 
                        () -> { 
                        Relation created = create(relationDescription, landscape);
                        affected.add(created);
                        processed.add(created);
                        processLog.info(String.format(origin + ": Adding relation between %s and %s", created.getSource(), created.getTarget()));
                        changelog.addEntry(created, ProcessingChangelog.ChangeType.CREATED, null);
               });
````
And then you could extract two methods and obtain:
````
                Optional<Relation> optionalRelation = getRelation(relationDescription, landscape, origin);
                optionalRelation.ifPresentOrElse(
                        (relation) -> updateRelation(
                                changelog, processed, affected, relationDescription, relation
                        ), 
                        () -> createRelation(
                                landscape, changelog, processed, origin, affected, relationDescription
                        )
                );
````

Reply:
nice 👍